sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -0,0 +1,124 @@
+  private static final int DEFAULT_CAPACITY_PER_SHARD = 10_000;
+  private ExecutorService executorService;
+  private BlockingQueue<KinesisRecord> recordsQueue;
+  private Map<String, ShardRecordsIterator> shardIteratorsMap;
+  private final int queueCapacity;
+  private volatile boolean poolOpened = true;
+
+  ShardReadersPool(List<ShardRecordsIterator> iterators) {
+    this(iterators, DEFAULT_CAPACITY_PER_SHARD);
+  }
+
+  ShardReadersPool(List<ShardRecordsIterator> iterators, int queueCapacity) {
+    this.shardIteratorsMap = Maps
+        .uniqueIndex(iterators, new Function<ShardRecordsIterator, String>() {
+
+          @Override
+          public String apply(ShardRecordsIterator input) {
+            return input.getShardId();
+          }
+        });
+    this.queueCapacity = queueCapacity;
+  }
+
+  void start() {
+    executorService = Executors.newFixedThreadPool(shardIteratorsMap.size());
+    recordsQueue = new LinkedBlockingQueue<>(queueCapacity * shardIteratorsMap.size());
+    for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+      executorService.submit(new Runnable() {
+
+        @Override
+        public void run() {
+          readLoop(shardRecordsIterator);
+        }
+      });
+    }
+  }
+
+  private void readLoop(ShardRecordsIterator shardRecordsIterator) {
+    while (poolOpened) {
+      try {
+        List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
+        for (KinesisRecord kinesisRecord : kinesisRecords) {
+          recordsQueue.put(kinesisRecord);
+        }
+      } catch (TransientKinesisException e) {
+        LOG.warn("Transient exception occurred.", e);
+      } catch (InterruptedException e) {
+        LOG.warn("Thread was interrupted, finishing the read loop", e);
+        break;
+      } catch (Throwable e) {
+        LOG.error("Unexpected exception occurred", e);
+      }
+    }
+    LOG.info("Kinesis Shard read loop has finished");
+  }
+
+  CustomOptional<KinesisRecord> nextRecord() {
+    try {
+      KinesisRecord record = recordsQueue.poll(1, TimeUnit.SECONDS);
+      if (record == null) {
+        return CustomOptional.absent();
+      }
+      shardIteratorsMap.get(record.getShardId()).ackRecord(record);
+      return CustomOptional.of(record);
+    } catch (InterruptedException e) {
+      LOG.warn("Interrupted while waiting for KinesisRecord from the buffer");
+      return CustomOptional.absent();
+    }
+  }
+
+  void stop() {
+    LOG.info("Closing shard iterators pool");
+    poolOpened = false;
+    executorService.shutdownNow();
+    boolean isShutdown = false;
+    while (!isShutdown) {
+      try {
+        isShutdown = executorService.awaitTermination(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.error("Interrupted while waiting for the executor service to shutdown");
+        throw new RuntimeException(e);
```

Comment:
Not sure about throwing RuntimeException here as it may do some harm to the stopping mechanism of Beam. 

The best way would be to simply propagate the InterruptedException, but I know this is hard to do as it's checked exception.

Reply:
I recalled that I took this logic from KafkaIO connector. It also throws RuntimeException in such case, with "not expected" comment. So I guess the shutdown procedure is not meant to be interrupted by Beam.